### PR TITLE
[W12] [v1.5] Implemented Google Login Feature

### LINKED ISSUE #169: Email Bug
Application crashes whenever it opens a browser tab to send an email out.

<sub>[original: nus-cs2103-AY1718S2/pe-round1#293]</sub>

 Issue created by: @lohtianwei

### PR DESCRIPTION
**Pull request by KevinCJH**
- Implemented Google login into the app's built in browser so that cheeky users cannot close the authentication browser.
- Added googlelogin command which will open the google login page in the app's browser.

Fixes #169 #170 #176 #191 #193 #200

### CODE REVIEW

File: docs/team/KevinChin.adoc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+contactHeRo is a desktop contact management application
+made for Human Resource (HR) managers, specifically for
+recruiters in a company.
+
+== Summary of contributions
+
+* *Major enhancement*: added *the ability to send out email* in contactHeRo, using the Google API. Also provided a *framework to integrate Google API*.
+** What it does: allows the user to send email to their contact, using an in-app editor. To use the Google API to send email, users will have to login to Google. Our app provides a built-in browser to handle the user login and a framework to integrate Google API.
+** Justification: As we all know, recruiters will usually send out emails to potential employees to inform them of their job application progress or arranging for interview.
```
Comment (trafalgarandre) [inline docs/team/KevinChin.adoc]: interviews

File: docs/team/KevinChin.adoc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+This feature lets user take advantage of the fact that the information of the potential employees are already in the app, therefore it will be easier to draft an email than using other email applications.
```
Comment (trafalgarandre) [inline docs/team/KevinChin.adoc]: "...is already in the app"

File: docs/team/KevinChin.adoc [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+It also provides a better user experience as users do not need to switch to another application.
+** Highlights: This enhancement provides a framework for future feature which requires Google API, like Google Contacts, Google Drive or Google Calendar.
```
Comment (trafalgarandre) [inline docs/team/KevinChin.adoc]: features

File: docs/team/KevinChin.adoc [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ It required the programming ability to integrate third-party API and third-party code into our own app.
+ The implementation was challenging as you do not have control over the source code in the third-party API, which means you have to work your own code around it.
+** Credits: stackoverflow.com, googledevelopers.com
+
+* *Minor enhancement*: added the *ability to find by tag*, which allows the user to find their contact using the skills tag.
+** What it does: It allows user to search for a contact based on their skills tag.
```
Comment (trafalgarandre) [inline docs/team/KevinChin.adoc]: the user

File: docs/team/KevinChin.adoc [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+** Justification: Sometimes user might not remember the name of the person they have shortlisted.
```
Comment (trafalgarandre) [inline docs/team/KevinChin.adoc]: the user

File: docs/UserGuide.adoc [hunk 6]
```diff
@@ -390,32 +390,73 @@ The `redo` command fails as there are no `undo` commands executed previously.
...
+// tag::googlelogin[]
+=== Login to Google : `googlelogin`
+
+You will have to login to google in order to use features like
+emailing. This process is simple and fast like how you normally login
+using the web browser. +
+Format: `googlelogin`
+
+****
+* Opens up the `Google` tab. This is contactHeRo's built in browser for Google logins. Simply enter your email and password to login.
```
Comment (trafalgarandre) [inline docs/UserGuide.adoc]: built-in

File: docs/UserGuide.adoc [hunk 7]
```diff
@@ -390,32 +390,73 @@ The `redo` command fails as there are no `undo` commands executed previously.
...
+* You will have to login in order to use the `Email` command to send email.
```
Comment (trafalgarandre) [inline docs/UserGuide.adoc]: send your email

File: docs/UserGuide.adoc [hunk 8]
```diff
@@ -390,32 +390,73 @@ The `redo` command fails as there are no `undo` commands executed previously.
...
+
+ Please log in to Google.
++
+You should also see the login screen as shown below.
+
+.The Google log in page
+image::googlelogin1.PNG[width="600"]
+
+[NOTE]
+Please do not go to any other webpage in the `Google` tab after you have logged in.
```
Comment (trafalgarandre) [inline docs/UserGuide.adoc]: webpages

File: docs/UserGuide.adoc (hunk 7 above)
Comment (trafalgarandre) [reply docs/UserGuide.adoc]: check all the word email. some of them you can add "your"

File: docs/DeveloperGuide.adoc [hunk 9]
```diff
@@ -592,31 +592,80 @@ https://docs.oracle.com/javafx/2/api/javafx/scene/image/Image.html
...
+Below is the class diagram of the `GmailClient`:
 
-Firstly, `GoogleAuthorizationCodeFlow` is used to build and request for authorization from user.
-Then it saves the user authorization into `Credential`, which is saved into a local file so that they do not need to repeat the authorization step.
-This authorization is handled in the `GmailAuthentication` class.
+.Class Diagram of GmailClient and GoogleAuthentication
+image::GmailClassDiagram.png[width="600", align="center"]
 
-Next, a `GmailClient` is created using the `Credential` that was created by `GmailAuthentication`.
-The `GmailClient` class provides all the services that uses the Gmail API.
```
Comment (trafalgarandre) [inline docs/DeveloperGuide.adoc]: use

File: docs/DeveloperGuide.adoc [hunk 10]
```diff
@@ -592,31 +592,80 @@ https://docs.oracle.com/javafx/2/api/javafx/scene/image/Image.html
...
+This newly created `GmailClient` class will provide all the services that uses the Google API.
```
Comment (trafalgarandre) [inline docs/DeveloperGuide.adoc]: use

File: docs/DeveloperGuide.adoc [hunk 11]
```diff
@@ -592,31 +592,80 @@ https://docs.oracle.com/javafx/2/api/javafx/scene/image/Image.html
...
+        message = service.users().messages().send("me", message).execute();
+    }
 
 ==== Design consideration
+===== Aspect: How to login to Google
+* Alternative 1: Using `GoogleAuthorizationCodeFlow` to log in.
+** Pros: Very simple to implement and handle. `Token` and `Credential` are handled by the API.
+** Cons: It opens a default web browser, which users are able to close. `GoogleAuthorizationCodeFlow` will stall the
+main thread in order to wait for user to authenticate, if user close the browser, the app will crash.
+* Alternative 2 (current choice): Using built-in web browser to get the `Token` from the URL.
```
Comment (trafalgarandre) [inline docs/DeveloperGuide.adoc]: a built-in web...
you can put some article before the user

File: docs/team/KevinChin.adoc (hunk 1 above)
Comment (KevinCJH) [reply docs/team/KevinChin.adoc]: changed. thanks

File: docs/DeveloperGuide.adoc (hunk 9 above)
Comment (KevinCJH) [reply docs/DeveloperGuide.adoc]: @trafalgarandre which one you referring to

**Pull request merged**